elytron/WFLY-16306-keycloak-saml-feature-pack.adoc

```diff
@@ -0,0 +1,331 @@
+added native support for OpenID Connect (see https://issues.jboss.org/browse/WFLY-13706[WFLY-13706] / https://issues.jboss.org/browse/KEYCLOAK-14953[KEYCLOAK-14953]).
+Thus, the analysis for this task is quite similar to the analysis for https://issues.jboss.org/browse/WFLY-13706[WFLY-13706].
+
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-16306[WFLY-13706 - Galleon feature pack to install Keycloak adapters]
+* https://github.com/keycloak/keycloak/issues/12363[Issue #12363 - Galleon feature pack to install Keycloak Elytron SAML adapter]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1928[EAP7-1928]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+
+TBD.
+
+Note that since this feature is being developed across two different projects, we'll need engineers and QE
+from both the WildFly and Keycloak projects involved.
+
+* [ ] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+ * https://github.com/keycloak/keycloak
+ * https://github.com/wildfly
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+Note that the code for the Keycloak Elytron SAML adapter is going to be extracted from the https://github.com/keycloak/keycloak repo
+and moved to a separate repo. We don't have details yet on when this will happen. We'll likely need to add our feature pack to the
+existing Keycloak project initially and then it will get moved to the new project with the rest of the adapter code.
+
+A new Galleon feature pack will be added to the Keycloak project with the group ID, `org.keycloak`, and artifact ID, `keycloak-saml-adapter-galleon-pack`.
+This new feature pack will contain three layers called `keycloak-saml`, `keycloak-client-saml`, and `keycloak-client-saml-ejb`.
+
+When this feature pack and its layers are provisioned to a WildFly installation, the end result will mostly match the results
+of manually installing the Keycloak Elytron SAML adapter using the instructions linked above.
+
+In particular, when the `keycloak-saml` layer is provisioned, the following steps will occur:
+
+ # Install a set of modules under `modules/system/add-ons`
+ # Register the `org.keycloak.keycloak-saml-adapter-subsystem` extension.
+ # Add an empty `urn:jboss:domain:keycloak:1.4` subsystem.
+
+The `keycloak-client-saml` layer will depend on the `keycloak-saml` layer and when provisioned, the following additional steps will occur:
+
+ # Add the required resources to the `elytron` subsystem to result in an `http-authentication-factory` and `security-domain` supporting SAML being available.
+ # Update the `application-security-domain` mapping in the `undertow` subsystem to map the default security domain to the new `http-authentication-factory`.
+
+The `keycloak-client-saml-ejb` layer will depend on the `keycloak-client-saml` layer and when provisioned, the following additional steps will occur:
```

Comment:
@jfdenise @OndrejKotek I noticed that the existing SAML adapter CLI scripts and the existing feature pack for the Keycloak Elytron OIDC adapter only added an `application-security-domain` mapping in the Undertow subsystem.

However, the OIDC adapter CLI scripts also added an `application-security-domain` mapping in the EJB subsystem.

So I've introduced an additional `keycloak-client-saml-ejb` layer here so that users who do want to propagate their identities to EJBs can make use of this. Note that I didn't include this in the `keycloak-client-saml` layer instead because that would require a dependency on the `ejb-lite` layer to be added to the `keycloak-client-saml` layer. For users with a configuration that doesn't already include the EJB subsystem, provisioning this layer would then cause a lot of extra configuration to be added unnecessarily. So I think it's better to have this in the separate `keycloak-client-saml-ejb` layer.

Reply:
@fjuma when using `keycloak-client-saml-ejb` is it necessary to specify `keycloak-client-saml` in the list of layers? same question for `keycloak-client-saml` and `keycloak-saml` .... I suppose each layer is an extension of the preceding and we don't have to specify them all, is that correct?

Reply:
@tommaso-borgato Right, you don't have to specify them all. So, as an example, you'd just need to specify `keycloak-saml-client-ejb`.